src/wasm/asm-wasm-builder.cc

```diff
@@ -434,6 +434,91 @@ class AsmWasmBuilderImpl : public AstVisitor {
     RECURSE(Visit(expr->expression()));
   }
 
+  template<typename T>
+  bool TestBinaryOperation(BinaryOperation* expr, Token::Value op, T val) {
+    if (expr->op() == op && expr->right()->IsLiteral()) {
+      Literal* right = expr->right()->AsLiteral();
+      if (static_cast<T>(right->raw_value()->AsNumber()) == val) {
```

Comment:
We have to be super careful here. This a C++ static cast, which has undefined semantics when converting floating point numbers out of range to integers, so this test is somewhat unreliable. Is there a way to also check the type (i.e. int32)? Or maybe we need to two separate functions, one for double and one for int32.